heltour/tournament/admin.py

```diff
@@ -45,19 +47,129 @@ class _BaseAdmin(VersionAdmin):
+
+    def get_queryset(self, request):
+        result = super(_BaseAdmin, self).get_queryset(request)
+        if self.allow_all_staff or self.has_assigned_perm(request.user, 'change'):
+            return result
+        if self.league_id_field is None:
+            return result.none()
+        return result.filter(**{self.league_id_field + '__in': self.authorized_leagues(request.user)})
+
+    def formfield_for_foreignkey(self, db_field, request, **kwargs):
+        kwargs['queryset'] = admin.site._registry[db_field.related_model].get_queryset(request)
+        return super(_BaseAdmin, self).formfield_for_foreignkey(db_field, request, **kwargs)
+
+    def has_add_permission(self, request):
+        if self.allow_all_staff or self.has_assigned_perm(request.user, 'add'):
+            return True
+        return self.has_league_perm(request.user, 'add', None)
+
+    def has_change_permission(self, request, obj=None):
+        if self.allow_all_staff or self.has_assigned_perm(request.user, 'change'):
+            return True
+        return self.has_league_perm(request.user, 'change', obj)
+
+    def has_delete_permission(self, request, obj=None):
+        if self.allow_all_staff or self.has_assigned_perm(request.user, 'delete'):
+            return True
+        return self.has_league_perm(request.user, 'delete', obj)
+
+    def get_form(self, request, obj=None, **kwargs):
+        form = super(_BaseAdmin, self).get_form(request, obj, **kwargs)
+
+        def clean(form):
+            super(ModelForm, form).clean()
+            self.clean_form(request, form)
+
+        form.clean = clean
+        return form
+
+    def clean_form(self, request, form):
+        if self.allow_all_staff:
+            return
+        if form.instance.pk is None and self.has_assigned_perm(request.user, 'add'):
+            return
+        if form.instance.pk is not None and self.has_assigned_perm(request.user, 'change'):
+            return
+        if self.league_id_field is None:
+            raise ValidationError('No permission to save this object')
+        # Since we have cleaned_data dict instead of a model instance, we have to
+        # pre-process the league id access a bit
+        parts = self.league_id_field.split('__', 1)
+        if len(parts) == 1:
+            if parts[0] == 'id':
+                league_id = form.cleaned_data['id']
+            else:
+                if parts[0][-3:] != '_id':
+                    raise ValueError('Invalid league id field on modeladmin')
+                league_id = form.cleaned_data[parts[0][:-3]].id
+        else:
+            league_id = getnestedattr(form.cleaned_data[parts[0]], parts[1])
+        if league_id not in self.authorized_leagues(request.user):
+            raise ValidationError('No permission to save objects for this league')
+
+    def authorized_leagues(self, user):
+        return [lm.league_id for lm in LeagueModerator.objects.filter(player__lichess_username__iexact=user.username)]
```

Comment:
You should use the `.values` query here if you're only getting IDs. It will be a fair bit faster